Deduplicate ensureValidMerge check

luyuncheng opened the pull request:
when there is a call in IndexWrier#maybeMerge->updatePendingMerges->registerMerge

i think the `ensureValidMerge` is duplicated check  which do the same logic in the code:

https://github.com/apache/lucene/blob/bed07c6b0242646306470df9265469f6a811b5b2/lucene/core/src/java/org/apache/lucene/index/IndexWriter.java#L4766-L4777

and when check error, it return false early.

luyuncheng committed:
deduplicate check

mikemccand commented:
Good catch!  Thanks @luyuncheng.

Except, the `ensureValidMerge` would throw an exception in this case -- can we fix the duplicated code to throw that same exception?

luyuncheng committed:
add exceptin in registerMerge && fixed test

luyuncheng commented:
> Except, the `ensureValidMerge` would throw an exception in this case -- can we fix the duplicated code to throw that same exception?

@mikemccand thanks for your advice,  i added throw exception in https://github.com/apache/lucene/pull/12237/commits/d14f5d4a254b425d2b4e3e14e71b4510b9fec818 and fixed test `testLiveMaxMergeCount` when throw the same exception in iterator

```
MergePolicy selected a segment (_2e) that is not in the current index _0(10.0.0)
```
